string_registers.c

```diff
@@ -0,0 +1,119 @@
+/*
+  string_registers.c - get/set string register value by id or name
+
+  Part of grblHAL
+
+  Copyright (c) 2024-2025 Stig-Rune Skansgård
+
+  grblHAL is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  grblHAL is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with grblHAL. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "hal.h"
+
+#if STRING_REGISTERS_ENABLE
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#include "system.h"
+#include "settings.h"
+#include "ngc_params.h"
+#include "string_registers.h"
+
+#ifndef MAX_SR_LENGTH
+#define MAX_SR_LENGTH 50
+#endif
+typedef struct string_register {
```

Comment:
If you move the value array to the end and change it to a single character (value[1]) and malloc the struct size plus the string length then you may save memory and lift (or increase) the maximum string length. Max. block (line) length in gcode is 256, this limit should perhaps be adhered to - remember plugin code may use the API for string handling directly.

Reply:
Wouldn't malloc'ing dynamicly like that cause issues if you modify a string-register with a longer value than it was originally malloc'ed with?

Reply:
Use realloc instead and check for increased length.

Reply:
huh, I thought an array of structs in C would be lined up one after another in memory, so I assumed each string_register had to be the same size 🤔 😅

Reply:
an array yes, but not a linked list - the last element in the struct can be any length since the struct is dynamically allocated and the next pointer points to the next element wherever that has been allocated...

Reply:
lol, of course you are correct. I somehow skipped over the fact that it is a linked list and not an array! 🤦 😂

Reply:
> I have to update the pointer to "next" in the previous item when I realloc an existing register, right?

Only if realloc cannot extend the current register allocation. But easier to just assume it has been moved.
And there is no need to use malloc at all for the initial allocation, if you pass a null pointer to realloc it behaves like malloc.